test/test_mpi_pack.c

```diff
@@ -0,0 +1,185 @@
+                        double startvalue)
+{
+  message->type = type;
+  message->values = SC_ALLOC (double, num_values[type]);
+  for (int i = 0; i < num_values[type]; i++) {
+    message->values[i] = startvalue * i;
+  }
+}
+
+void
+test_message_destroy (test_message_t *message)
+{
+  SC_FREE (message->values);
+}
+
+int
+test_message_equal (test_message_t *message1, test_message_t *message2)
+{
+  if (message1->type != message2->type) {
+    return 0;
+  }
+  for (int ivalue = 0; ivalue < num_values[message1->type]; ivalue++) {
+    if (message1->values[ivalue] != message2->values[ivalue]) {
+      return 0;
+    }
+  }
+  return 1;
+}
+
+/* Pack several instances of test messages into contiguous memory */
+int
+test_message_MPI_Pack (const test_message_t *messages, int incount,
+                       void *outbuf, int outsize, int *position,
+                       sc_MPI_Comm comm)
+{
+  for (int imessage = 0; imessage < incount; imessage++) {
+    sc_MPI_Pack (&(messages[imessage].type), 1, sc_MPI_INT8_T, outbuf,
+                 outsize, position, comm);
+    sc_MPI_Pack (messages[imessage].values,
+                 num_values[messages[imessage].type], sc_MPI_DOUBLE, outbuf,
+                 outsize, position, comm);
+  }
+  return 0;
+}
+
+/* Unpack contiguous memory into several instances of the same datatype */
+int
+test_message_MPI_Unpack (const void *inbuf, int insize,
+                         int *position, test_message_t *messages,
+                         int outcount, sc_MPI_Comm comm)
+{
+  for (int imessage = 0; imessage < outcount; imessage++) {
+    sc_MPI_Unpack (inbuf, insize, position, &(messages[imessage].type), 1,
+                   sc_MPI_INT8_T, comm);
+    messages[imessage].values =
+      SC_ALLOC (double, num_values[messages[imessage].type]);
+    sc_MPI_Unpack (inbuf, insize, position, messages[imessage].values,
+                   num_values[messages[imessage].type], sc_MPI_DOUBLE, comm);
+  }
+  return 0;
+}
+
+/* Determine how much space in bytes is needed to pack several test messages */
+int
+test_message_MPI_Pack_size (int incount, const test_message_t *messages,
+                            sc_MPI_Comm comm, int *size)
+{
+  *size = 0;
+  for (int imessage = 0; imessage < incount; imessage++) {
+    int                 pack_size;
+    int                 single_message_size = 0;
+
+    sc_MPI_Pack_size (1, sc_MPI_INT8_T, comm, &pack_size);
+    single_message_size += pack_size;
+    sc_MPI_Pack_size (1, sc_MPI_DOUBLE, comm, &pack_size);
+    single_message_size += num_values[messages[imessage].type] * pack_size;
+    *size += single_message_size;
+  }
+  return 0;
+}
+
+int
+main (int argc, char **argv)
+{
+  int                 mpiret;
+  sc_MPI_Comm         mpicomm;
+
+  /* standard initialization */
+  mpiret = sc_MPI_Init (&argc, &argv);
+  SC_CHECK_MPI (mpiret);
+  mpicomm = sc_MPI_COMM_WORLD;
+  sc_init (mpicomm, 1, 1, NULL, SC_LP_DEFAULT);
+
+  int                 num_test_messages = 5;
```

Comment:
These and further declarations please move to the top before any code is executed.